lib/redis.js

```diff
@@ -32,63 +33,77 @@ module.exports = class Redis {
+      logger.error('Redis error:', err);
+      this.disconnected = true;
+    });
   }
 
-  hostName() {
-    if (this._host) {
-      return 'redis://' + this._host.replace(/^.*\/\//, '').replace(/:.*$/, '');
+  static buildConn(redisUrl, maxRetries) {
+    const opts = {
+      lazyConnect: true,
+      maxRetriesPerRequest: maxRetries,
+      retryStrategy: times => {
+        if (times <= maxRetries) {
+          return Math.min(times * 50, 2000);
+        } else {
+          this.disconnected = true;
+          return null;
+        }
+      }
+    };
+
+    // optionally run in cluster mode
+    if (redisUrl.match(/^cluster:\/\/.+:[0-9]+$/)) {
+      const parts = redisUrl.replace(/^cluster:\/\//, '').split(':');
+      const config = [{host: parts[0], port: parts[1]}];
+      return new IORedis.Cluster(config, {lazyConnect: true, clusterRetryStrategy: opts.retryStrategy, redisOptions: opts});
     } else {
-      return null;
+      return new IORedis(redisUrl, opts);
     }
   }
 
-  connect() {
-    return new Promise((resolve, reject) => {
-      let hostKey = (this._host || '').match(/^redis:/) ? 'url' : 'host';
-      this._client = redis.createClient({
-        connect_timeout: 5000, // don't wait forever
-        [hostKey]: process.env.REDIS_HOST,
-        retry_strategy: (opts) => {
-          if (opts.error && opts.error.code === 'ECONNREFUSED') {
-            logger.error('Redis client error: ECONNREFUSED');
-            reject(opts.error);
-          }
-          if (opts.attempt < 10) {
-            return 100;
-          } else if (opts.error) {
-            reject(opts.error);
-          } else {
-            reject(new Error(`Failed to connect to redis after ${opts.attempt} attempts`));
-          }
-          return undefined;
-        }
-      });
-      this._client.on('error', err => reject(err));
-      this._client.on('connect', () => resolve(this._client));
-    });
+  get connected() {
+    return !this.disconnected;
   }
 
-  disconnect() {
-    return new Promise((resolve, reject) => {
-      this._client.on('error', err => reject(err));
-      this._client.on('end', () => resolve());
-      this._client.quit();
-    });
+  hostName() {
+    if (this.url && !this.url.match(/^(redis)|(cluster):\/\//)) {
+      return 'redis://' + this.url.replace(/^.*\/\//, '').replace(/:[0-9]+$/, '');
+    } else {
+      return (this.url || '').replace(/:[0-9]+$/, '') || null;
+    }
+  }
+
+  async disconnect() {
+    try {
+      this.disconnected = true;
+      await this.conn.quit();
+      return true;
+    } catch (err) {
+      logger.error('Redis client error:', err);
+      return false;
+    }
   }
 
   increment(key, field, count) {
-    return new Promise((resolve, reject) => {
-      this._client.hincrby(key, field, count, (err, reply) => resolve(reply));
-    });
+    if (this.disconnected) {
+      this.conn.connect().catch(err => null);
+    }
+    return this.conn.hincrby(key, field, count);
```

Comment:
Are these on the main thread?  Maybe errors are caught and percolated as part of the Input class.

Reply:
Ahh up in RedisActuals#insert

Reply:
Yeah, basically in a Lambda (unlike the `dovetail.prx.org` node app) you really have to wait for ever redis call to complete.  Or else the lambda will hang until all redis connections are done/closed.